instrumentation/CONTRIBUTING.md

```diff
@@ -0,0 +1,483 @@
+Instrumentation libraries should be as lightweight as possible and must:
+
+* Rely on `rubocop-performance` linters to catch performance issues
+* Consider using [microbenchmarks](https://github.com/evanphx/benchmark-ips) and [profiling](https://ruby-prof.github.io/) to address any possible performance issues
+* Provide minimal solutions and code paths
+
+#### Provide minimal solutions and code paths
+
+Instrumentations should have the minimal amount of code necessary to provide useful insights to our users. It may sound contrary to good engineering practices, but you must avoid adding lots of small methods, classes, and objects when instrumenting a library.
+
+Though often easier to maintain and reason about; small and well factored code adds overhead to the library you are instrumenting resulting in performance degradation due to unnecessary object allocations, method dispatching, and other performance overhead.
+
+It will also contribute to building large backtraces that makes it more difficult for our end users to understand the essential parts of exception reports application. That will in turn likely result in additional filtering logic in the their application to avoid reporting unnecessary stack frames.
+
+In cases when code uses monkey patching, it runs the risk of _adding_ methods that conflict with the internal implementatation of the library and may result in unexpected behavior and bugs.
+
+Avoid instrumenting _every_ method in a library and instead focus on the methods the provide the _most_ insights into what typically causes performance problems for applications, e.g. I/O and network calls. The use case for this type of low level granularity fails under the purview of profiling.
+
+In the near future, [OTel Profiling](https://opentelemetry.io/blog/2024/profiling/) will provide users an even deeper understanding of what is happening in their applications at a more granular level.
+
+Here are some examples of performance fixes that reduced object allocations and method dispatching:
+
+* <https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/867>
+* <https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/723>
+* <https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/665>
+* <https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/642>
+* <https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/207>
+* <https://github.com/open-telemetry/opentelemetry-ruby-contrib/pull/232>
+
+#### Avoid Adding Custom Extensions
+
+Though your instrumentation may accept configurations options to customize the output, you should consider that the more options you add, the more complexity you will have to manage.
+
+You should _avoid_ adding options that allow custom code blocks to be executed as part of the instrumentation. It is often difficult to predict error modes and the performance impact custom code will have on your instrumentation, which in turn will impact the service being instrumented.
```

Comment:
```suggestion
You should _avoid_ adding options that allow custom code blocks (`type: :callable`) to be executed as part of the instrumentation. It is often difficult to predict error modes and the performance impact custom code will have on your instrumentation, which in turn will impact the service being instrumented.
```

Reply:
Now that I am reading this again, I think it's best to change the language and say `You must` as opposed to `You should`